server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java

```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
     @Override
     public List<IdentityZone> retrieveAll() {
         try {
-            String systemTenant = this._idmClient.getSystemTenant();
             ArrayList<IdentityZone> list = new ArrayList<IdentityZone>();
             for (String tenant : this._idmClient.getAllTenants()) {
-                list.add(identityZoneForTenant(this._idmClient.getTenant(tenant),
-                        systemTenant.equalsIgnoreCase(tenant)));
+                list.add(getIdentityZoneFromTenant(tenant));
             }
-
+            
             return list;
         } catch (Exception ex) {
             throw new IllegalStateException("Failed to retrieve zones.", ex);
         }
     }
+    
+    private IdentityZone identityZoneForTenant(String tenantName) throws Exception {
+        IdentityZone zone = new IdentityZone();
+        Tenant t = this._idmClient.getTenant(tenantName);
+        
+        zone.setId(t.getName());
+        zone.setName(t.getName());
+        zone.setSubdomain(t.getName());
+        zone.setDescription(t._longName);
+        zone.setVersion(1);
+        
+        return zone;
+    }
+    
+    private void configureIdentityZone(String tenantName, IdentityZone zone) throws Exception {
+        IdentityZoneConfiguration config = this._configs.get(tenantName);
+        if (config == null) {
+            config = new IdentityZoneConfiguration();
+            this._configs.put(tenantName, config);
+        }
+        
+        Map<String, String> keys = new HashMap<String, String>();
+        PrivateKey privateKey = _idmClient.getTenantPrivateKey(tenantName);
+        StringWriter writer = new StringWriter();
+        PEMWriter pemWriter = new PEMWriter(writer);
+        pemWriter.writeObject(privateKey);
+        pemWriter.close();
+        
+        keys.put("key-id-1", writer.toString());
+        config.getTokenPolicy().setKeys(keys);
+        config.getTokenPolicy().setActiveKeyId("key-id-1");
+        config.getTokenPolicy().setAccessTokenValidity(
+                Ints.saturatedCast(_idmClient.getMaximumBearerTokenLifetime(tenantName)));
+        config.getTokenPolicy().setRefreshTokenValidity(
+                Ints.saturatedCast(_idmClient.getMaximumBearerRefreshTokenLifetime(tenantName)));
+        
+        zone.setConfig(config);
+    }
 
-    private IdentityZone identityZoneForTenant(Tenant t, boolean systemTenant) {
-        IdentityZone identityZone = null;
-        if (systemTenant) {
-            identityZone = IdentityZone.getUaa();
+    private IdentityZone getIdentityZoneFromId(String id) throws Exception {
+        String tenantName = id;
+        IdentityZone zone = null;
+        
+        if (id.equalsIgnoreCase(IdentityZone.getUaa().getId())) {
```

Comment:
The trouble is that this structure is being used to do two different things: not only are we retrieving the true system tenant name (based on the fact we were in the default UAA zone), but we're fetching the default `IdentityZone` as well.

To use that method, I have to do two things: I have to fetch the system tenant's name (which is kind of pointless when we're not in the default zone) and I have to then compare the output of `#getTenantName()` against the system tenant to determine which zone I pull. For example:

    String systemTenant = this._idmClient.getSystemTenant();
    String tenantName = VmidentityUtils.getTenantName(id, systemTenant);
    if (tenantName.equalsIgnoreCase(systemTenant)) {
         zone = IdentityZone.getUaa();
    } else {
        zone = identityZoneForTenant(tenantName);
    }

Doing that both creates an unnecessary remote IDM call in the non-UAA zone (e.g. a named tenant) and creates a second comparison that's functionally the same as comparing the zone against the default UAA zone. I also can't remove the remote IDM call at this point, because I'll wind up needing the name of the system tenant when it comes time to fetch the `IdentityZone`.